### ISSUE TITLE
Using angularfire2 with this starter

### ISSUE THREAD

**Issue by beckmeindia** [angular/universal-starter#127]
Hey,
     I am using Angular2 webpack with Firebase for developing my application. Hence I am using this starter to build the entire application. 

I searched through documentation for using firebase or angularfire2 with this but could not find any. The present angularfire2 docs use angular-cli & don't work with this.

Could you please help?

Thanks,
Rahul

**Comment by manekinekko** [comment]
@beckmeindia what are you trying to accomplish? This starter is for Angular Universal not AngularFire.

**Comment by joaogarin** [comment]
@beckmeindia I think you might be in the wrong repo? https://github.com/angular/angularfire2
Documentation is on https://github.com/angular/angularfire2/blob/master/docs/1-install-and-setup.md

**Comment by beckmeindia** [comment]
@joaogarin thank you. I was referring to the wrong docs.

**Comment by beckmeindia** [comment]
Hey, 
    I am trying to integrate angularfire2 with this starter. Apparently I keep getting an error AngularFire not found, as it is running on the server side. The setup seems to be fine for webpack. I don't know how to bypass this problem of server-side and make it work with this starter. 

Could any of you give ideas about a workaround for it?

I have attached the repository here: https://github.com/beckmeindia/universal-angularfire2

I have also tried to address this issue here: https://github.com/r-park/todo-angular2-firebase/issues/77#issuecomment-234967193

Thanks,
Rahul

**Comment by manekinekko** [comment]
@beckmeindia I am gonna take a look at your issue tonight. I'll let you know.

**Comment by beckmeindia** [comment]
Thank you so much Wassim

**Comment by manekinekko** [comment]
@beckmeindia I did some diggings. Here are my tldr: We simply need to add support for AngularFire2.

Here are more details:

One of the issues is coming from the `FirebaseSdkAuthBackend` implementation which requests `this._fbAuth = _fbApp.auth();`. However, `_fbApp.auth()` doesn't seem to work (??); and then `this.-fbAuth` is `NULL` all the way down.

Here are the steps to reproduce this issue:

1. instead of providing `FIREBASE_PROVIDERS`,
1. provide these 2 deps:

```
COMMON_PROVIDERS,
{provide: AuthBackend, useClass: MockedAuthBackend}
```
And then provide a custom implementation with stubbed methods, something like:

```
class MockedAuthBackend {
   onAuth() {
     return null;
  }
}
```

Following the base parent class:

```
export declare abstract class AuthBackend {
    abstract authWithCustomToken(token: string): Promise<FirebaseAuthState>;
    abstract authAnonymously(options?: any): Promise<FirebaseAuthState>;
    abstract authWithPassword(credentials: EmailPasswordCredentials): Promise<FirebaseAuthState>;
    abstract authWithOAuthPopup(provider: AuthProviders, options?: any): Promise<firebase.auth.UserCredential>;
    abstract authWithOAuthRedirect(provider: AuthProviders, options?: any): Promise<void>;
    abstract authWithOAuthToken(credentialsObj: firebase.auth.AuthCredential, options?: any): Promise<FirebaseAuthState>;
    abstract onAuth(): Observable<FirebaseAuthState>;
    abstract getAuth(): FirebaseAuthState;
    abstract unauth(): void;
    abstract createUser(credentials: EmailPasswordCredentials): Promise<FirebaseAuthState>;
    abstract getRedirectResult(): Observable<firebase.auth.UserCredential>;
}
```

But then you will end up with other errors, like this one:

![image](https://cloud.githubusercontent.com/assets/1699357/17158718/9ae61a86-5398-11e6-8aea-7bbbe3f90398.png)

So basically, we need a universal AuthBackend implementation.

Getting @davideast in the loop ^^

**Comment by beckmeindia** [comment]
@manekinekko  Hey Wassem thanks a lot for your time and debugging. Will look forward to changes in the universal Authbackend by @davideast and try to implement what you have said & see if it somehow maybe works.

Cheers,
Rahul

**Comment by manekinekko** [comment]
@beckmeindia sure

**Comment by beckmeindia** [comment]
hey @manekinekko & @davideast 
Any updates for the auth implementation?
Really excited to implement it on my website!

**Comment by MarkPieszak** [comment]
Others interested in this as well. https://github.com/aspnet/JavaScriptServices/issues/268
Hopefully we can get this in for 1.0 if it's not too involving?

**Comment by beckmeindia** [comment]
Hi, I observed that the new angular2-webpack-starter has a very simple approach for integrating Angularfire2 through this:
add 'firebase' to my tsconfig.json file.
make sure you install typescript 2.0.0 npm install typescript@2.0.0 --save
install angularfire with npm install angularfire2@next -- save
add "firebase" to types array in tsconfig in the module
{
  "compilerOptions": {
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": ["es6", "dom"],
    "mapRoot": "./",
    "module": "es6",
    "moduleResolution": "node",
    "outDir": "../dist/out-tsc",
    "sourceMap": true,
    "target": "es5",
    "typeRoots": [
      "../node_modules/@types"
    ],
    "types": [
      "jasmine",
      "firebase"
    ]
  }
}

I have observed that the universal-starter uses an earlier version of angular2-webpack-starter. 
Could you suggest a way to upgrade to the new angular2-webpack? 
@davideast @MarkPieszak @manekinekko

**Comment by beckmeindia** [comment]
Hey @manekinekko 
   Does the new update to rc6 solves the problem we were facing with Auth earlier? 

Excited to hear about it

**Comment by jeffwhelpley** [comment]
@manekinekko what is the current status of this issue? I am just catching up on open problems and it seems like there was some back and forth on this as we were playing catchup with the angular release. Can you check again on this when you get a chance and let me know?

**Comment by beckmeindia** [comment]
I just checked with the updated versions of angularfire2 and angular-universal-nodejs.

It works perfectly fine with all the observables and authentication.  So it's resolved.

Thank you so much guys!
@manekinekko @jeffwhelpley @MarkPieszak

**Comment by doczoidberg** [comment]
is anywhere a starter project for angular universal with angularfire? Is am still struggling with it

**Comment by joaogarin** [comment]
Does universal play well with angularfire2? I see from https://github.com/angular/angularfire2/issues/191 that it looks like some custom code is needed and using the firebase nodejs version on the server..is that so?

**Comment by manekinekko** [comment]
@doczoidberg @joaogarin can you share some logs or error messages?

**Comment by joaogarin** [comment]
I didn't actually try it yet..but I am thinking about using firebase, and if that is a good idea at this point..I will definitely be using universal.

I will try to get some tests running and see what happens. I will report back with what I find.

**Comment by joaogarin** [comment]
Hello,

Just out of curiosity I tried putting angularfire2 into this repository and I get the following error : 

```
Error: Unexpected value '[object Object]' imported by the module 'AppModule'
    at /Library/WebServer/Documents/universal-starter/dist/server/index.js:20332
```

**Comment by samvloeberghs** [comment]
@joaogarin instead of importing AngularFireModule in the imports you could provide a substitute for AngularFire service:

class AngularFireServer{
  constructor(){}
  ...
}
export function createAngularFireServer(){
  return new AngularFireServer();
}

providers: [
    ...
    {provide: 'req', useFactory: getRequest},
    {provide: 'res', useFactory: getResponse},
    {provide: AngularFire, useFactory: createAngularFireServer},
    ...
]

I mainly use AngularFire / Firebase for Auth.
In this AngularFireServer I use the req to get a cookie and do authentication with the JWT in it..

**Comment by joaogarin** [comment]
Yeah thats what I am doing now, essencialy have a provider for browser and server =) bit of DI swap magic and it works :+1:

**Comment by michaelwaihenya** [comment]
+1. Any update on the possibility of using Universal with angularfire2 4.0.0-rc.0? There is a reference to window in database.js that does not seem to allow this.

**Comment by samvloeberghs** [comment]
I didn't tries with dB, but auth seems to work fine
